[JIT] Update docs for recently added features

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45234 [JIT] Add recursive scripting for class type module attributes
* #45262 [JIT] Add API for ignoring arbitrary module attributes
* #45261 [JIT] Enable @unused syntax for ignoring properties
* #45233 [JIT] Add test for ignored class type property
* **#45232 [JIT] Update docs for recently added features**

**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

Differential Revision: [D23971884](https://our.internmc.facebook.com/intern/diff/D23971884)

Commit:
[JIT] Update docs for recently added features

**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

SplitInfinity commented:
Where's a good place to mention `property` support? They're not part of the core Python language so `jit_python_reference.rst` is not the right place to put documentation about them. A new section in `jit_language_reference.rst`, maybe?

SplitInfinity commented on docs/source/jit_unsupported.rst:
```diff
@@ -87,6 +87,5 @@ we suggest using :meth:`torch.jit.trace`.
   * :class:`torch.nn.RNN`
   * :class:`torch.nn.AdaptiveLogSoftmaxWithLoss`
   * :class:`torch.autograd.Function`
-  * :class:`torch.autograd.no_grad`
```
Is there a good place to mention that `record_function` is supported now?

Commit:
Update on "[JIT] Update docs for recently added features"


**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

SplitInfinity commented on docs/source/jit_language_reference.rst:
```diff
@@ -346,6 +348,73 @@ like any other TorchScript type:
+::
+    from enum import Enum
+
+    global IntEnum
```
Fixed.

SplitInfinity commented on docs/source/jit_language_reference.rst:
```diff
@@ -346,6 +348,73 @@ like any other TorchScript type:
+After an enum is defined, it can be used in both TorchScript and Python interchangeably
+like any other TorchScript type:
+
+::
```
Fixed.

Commit:
Update on "[JIT] Update docs for recently added features"


**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

eellison commented on docs/source/jit_language_reference.rst:
```diff
@@ -346,6 +348,74 @@ like any other TorchScript type:
+
+    TorchScript enum support is experimental..
+
+Python enums can be used in TorchScript without any extra annotation or code:
```
Personally I would cut down on this section a bit - it's pretty long for what is an easy & simple feature (enums).   see how short the named tuple section is

SplitInfinity replied:
What do you think I should cut out? Most of it is code samples.

eellison reviewed (commented):
Looks good, a couple comments

gmagogsfm commented on docs/source/jit_language_reference.rst:
```diff
@@ -346,6 +348,74 @@ like any other TorchScript type:
+
+.. warning::
+
+    TorchScript enum support is experimental..
```
lol

SplitInfinity commented on docs/source/jit_language_reference.rst:
```diff
@@ -346,6 +348,74 @@ like any other TorchScript type:
 
 
+.. _TorchScript Enums:
+.. _TorchScript Enums:
```
I copied over the class types section, I don't know how to write `.rst` lol

eellison commented on docs/source/jit_language_reference.rst:
```diff
@@ -346,6 +348,74 @@ like any other TorchScript type:
+
+    @torch.jit.script
+    def return_enum(cond: bool):
+        if cond:
+            return Color.RED
+        else:
+            return Color.GREEN
```
And then maybe include this example where you access the field name `Color.RED` or make it part of the other function

eellison reviewed (approved):
Meant to accept - these are just suggestions though, it looks good obviously feel free to keep whatever

Commit:
Update on "[JIT] Update docs for recently added features"


**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

Commit:
Update on "[JIT] Update docs for recently added features"


**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

Commit:
Update on "[JIT] Update docs for recently added features"


**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

Commit:
Update on "[JIT] Update docs for recently added features"


**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

Commit:
Update on "[JIT] Update docs for recently added features"


**Summary**
This commit updates the TorchScript language reference to include
documentation on recently-added TorchScript enums. It also removed
`torch.no_grad` from the list of known unsupported `torch` modules and
classes because it is now supported.

**Test Plan**
Continuous integration.

[ghstack-poisoned]
